File: projects/erlang/self_balance_robot/src/self_balance_robot.erl
-module(self_balance_robot).

-export([start/0, handle_PID/5]).

-include("ledc.hrl").

% Macro for Timer and PWM channel
-define(LEDC_HS_TIMER, ?LEDC_TIMER_0).
-define(LEDC_HS_MODE, ?LEDC_HIGH_SPEED_MODE).
-define(LEDC_HS_CH0_GPIO, 19).
-define(LEDC_HS_CH0_CHANNEL, ?LEDC_CHANNEL_0).
-define(LEDC_HS_CH1_GPIO, 18).
-define(LEDC_HS_CH1_CHANNEL, ?LEDC_CHANNEL_1).

% This macro for (2^?LEDC_TIMER_13_BIT - 1) / 100
-define(CAL_BIT_DIV_100,  81.91).
% This macro for 100 / (2^LEDC_TIMER_13_BIT - 1)
-define(CAL_100_DIV_BIT, 0.012208521548040533).

-define(ALPHA, 0.988).
-define(SAMPLE_TIME_MS, 5).
-define(SAMPLE_TIME_S, 0.01).
-define(DIV_SAMPLE_TIME_S, 100).
-define(TARGET_ANGLE, -2).

-define(GPIO_SCL, 22).
-define(GPIO_SDA, 21).
-define(GPIO_LED, 2).

-define(MOTOR_LEFT_1, 16).
-define(MOTOR_LEFT_2, 4).
-define(MOTOR_RIGHT_1, 5).
-define(MOTOR_RIGHT_2, 23).

-define(MPU9250_ADDR, 16#68).
-define(ACCEL_YOUT_H, 16#3D).
-define(ACC_CONFIG_ADDR, 16#1C).
-define(GYRO_CONFIG_ADDR, 16#1B).

-define(I2C_BASE_FREQ, 400000).

% Macro for config Gyro
-define(GYRO_FULL_SCALE_250_DPS, 16#00).
-define(GYRO_FULL_SCALE_500_DPS, 16#08).
-define(GYRO_FULL_SCALE_1000_DPS, 16#10).
-define(GYRO_FULL_SCALE_2000_DPS, 16#18).

% Macro for config Accelerometer
-define(ACC_FULL_SCALE_2_G, 16#00).
-define(ACC_FULL_SCALE_4_G, 16#08).
-define(ACC_FULL_SCALE_8_G, 16#10).
-define(ACC_FULL_SCALE_16_G, 16#18).


% Scale with config Gyro: 2000DPS and Acc: 16G
-define(TEMP_OFFSET, 0).
-define(TEMP_SENS, 0.003115264797507788). % Equal to 1/321
-define(ACC_SCALE, 6.103515625e-5).         % Equal to 1/16384
-define(GYRO_SCALE, 0.007633587786259542). % Equal to 1/131
-define(RADIAN_TO_DEGREE, 57.2957795).

% Number of byte will read
-define(NUM_OF_BYTE, 8).
-define(NUM_OF_TIMES, 200).
-define(DIV_TIMES, 0.02).

-define(ACC_OFFSET, 4.5).
-define(GYRO_OFFSET, -1.75).
-define(AY_OFF, -0.00247998047).
-define(AZ_OFF, 1.00272071).
-define(GX_OFF, 376.86).
-define(DUTY_EQUAL_ZERO, 5000).
-define(MAX_SPEED, 8191).

-record(mpu, {accy, accz, gyrox, temp}).
%-record(error, {accerror, gyroerror}).

start() ->
    erlang:system_flag(schedulers_online, 2),
    I2C = i2c_init(),
    mpu_config(I2C),
    gpio_init(),
    pwm_init(),
    % Error = #error{accerror = 0, gyroerror = 0},
    % ErrorData = read_N_times(I2C, Error, ?NUM_OF_TIMES),
    % io:format("~p~n", [ErrorData]),
    %gpio:digital_write(?GPIO_LED, high),
    process_init(I2C),
    io:format("INIT OK ~n"),
    loop().

loop() ->
    % Use for test, type code here
    receive
        {power, Power, Angle} ->
            NewPower = constrain(Power, -8191, 8191),
            DutyNoneRound = abs(NewPower),
            %Duty = map(DutyNoneRound, 0, 8191, ?DUTY_EQUAL_ZERO, 8191),

            if 
                abs(Angle) > 10 ->
                    Duty = map(DutyNoneRound, 0, 8191, 7300, 8191);
                true ->
                    Duty = map(DutyNoneRound, 0, 8191, 5000, 6000)
            end,

            io:format("Duty ~p~n", [Duty]),
            if
                abs(Angle) < 1 ->
                    set_counter_channel_1(0),
                    set_counter_channel_2(0);
                true ->
                    if 
                        NewPower =< 0 ->
                            car_backward(),
                            set_counter_channel_2(Duty),
                            set_counter_channel_1(Duty);    
                        true ->
                            car_forward(),
                            set_counter_channel_2(Duty),
                            set_counter_channel_1(Duty)
                    end
            end;
        {angle, Angle} ->
            io:format("Current Angle is: ~p ~n", [round(Angle)]);     
        {stop, Angle} ->
            io:format("Angle is: ~p ~n", [Angle]),
            car_stop(),
            set_counter_channel_1(0),
            set_counter_channel_2(0)
    end,
    loop().

% Read value from MPU 6500
read(I2C) ->
    {ok, Val} = mpu_read_data(I2C),
    % Extract bit with format <<AccY_Z:32, Temp:16, GyroX:16>>
    <<AccY:16/integer-signed, AccZ:16/integer-signed,
    Temp:16,
    GyroX:16/integer-signed>> = Val,

    AccDataY = (AccY - ?AY_OFF) * ?ACC_SCALE,
    AccDataZ = (AccZ - ?AZ_OFF) * ?ACC_SCALE,
    GyroDataX = (GyroX - ?GX_OFF) * ?GYRO_SCALE,
    TempData = get_temp_value(Temp),

    #mpu{accy = AccDataY, accz = AccDataZ, gyrox = GyroDataX, temp = TempData}.

get_temp_value(Temp) ->
    (Temp - ?TEMP_OFFSET) * ?TEMP_SENS + 21.

i2c_init() ->
    i2c:open([{scl, ?GPIO_SCL}, {sda, ?GPIO_SDA}, {clock_speed_hz, ?I2C_BASE_FREQ}]).

mpu_config(I2C) ->
    mpu_send_command(I2C, ?ACC_CONFIG_ADDR, ?ACC_FULL_SCALE_2_G),
    mpu_send_command(I2C, ?GYRO_CONFIG_ADDR, ?GYRO_FULL_SCALE_250_DPS).

mpu_send_command(I2C, Register, Command) ->
    i2c:begin_transmission(I2C, ?MPU9250_ADDR),
    i2c:write_byte(I2C, Register),
    i2c:write_byte(I2C, Command),
    i2c:end_transmission(I2C).

% Return bitstring, format: <<AccY_Z:32, Temp:16, GyroX:16>>
mpu_read_data(I2C) ->
    i2c:begin_transmission(I2C, ?MPU9250_ADDR),
    i2c:write_byte(I2C, ?ACCEL_YOUT_H),
    i2c:end_transmission(I2C),
    i2c:read_bytes(I2C, ?MPU9250_ADDR, ?NUM_OF_BYTE).


% GPIO and Car controll API Part

gpio_init() ->
    List = [?MOTOR_LEFT_1, ?MOTOR_LEFT_2, ?MOTOR_RIGHT_1, ?MOTOR_RIGHT_2, ?GPIO_LED],
    lists:foreach(fun(GPIO) -> gpio:set_pin_mode(GPIO, output) end, List),
    car_stop().

write_motor_left(Status1, Status2) ->
    gpio:digital_write(?MOTOR_LEFT_1 , Status1),
    gpio:digital_write(?MOTOR_LEFT_2, Status2).

write_motor_right(Status1, Status2) ->
    gpio:digital_write(?MOTOR_RIGHT_1 , Status1),
    gpio:digital_write(?MOTOR_RIGHT_2, Status2).    

car_stop() ->
    write_motor_left(low, low),
    write_motor_right(low, low).

car_forward() ->
    write_motor_left(low, high),
    write_motor_right(low, high).

car_backward() ->
    write_motor_left(high, low),
    write_motor_right(high, low).

% PWM Part

pwm_init() ->
    LEDCHSTimer = [
        {duty_resolution, ?LEDC_TIMER_13_BIT},
        {freq_hz, 5000},
        {speed_mode, ?LEDC_HS_MODE},
        {timer_num, ?LEDC_HS_TIMER}
    ],
    ok = ledc:timer_config(LEDCHSTimer),
    LEDCChannel_1 = [
                {channel, ?LEDC_HS_CH0_CHANNEL},
                {duty, 0},
                {gpio_num, ?LEDC_HS_CH0_GPIO},
                {speed_mode, ?LEDC_HS_MODE},
                {hpoint, 0},
                {timer_sel, ?LEDC_HS_TIMER}
            ],
    LEDCChannel_2 = [
                {channel, ?LEDC_HS_CH1_CHANNEL},
                {duty, 0},
                {gpio_num, ?LEDC_HS_CH1_GPIO},
                {speed_mode, ?LEDC_HS_MODE},
                {hpoint, 0},
                {timer_sel, ?LEDC_HS_TIMER}
            ],
    ok = ledc:channel_config(LEDCChannel_1),
    ok = ledc:channel_config(LEDCChannel_2),
    ok = ledc:fade_func_install(0).

set_counter_channel_1(DutyNum) ->
    SpeedMode = ?LEDC_HS_MODE,
    Channel = ?LEDC_HS_CH0_CHANNEL,
    Duty = round(DutyNum),
    ok = ledc:set_duty(SpeedMode, Channel, Duty),
    ok = ledc:update_duty(SpeedMode, Channel).

set_counter_channel_2(DutyNum) ->
    SpeedMode = ?LEDC_HS_MODE,
    Channel = ?LEDC_HS_CH1_CHANNEL,
    Duty = round(DutyNum),
    ok = ledc:set_duty(SpeedMode, Channel, Duty),
    ok = ledc:update_duty(SpeedMode, Channel).

% Create process to handle read sensor

process_init(I2C) ->
    spawn(?MODULE, handle_PID, [I2C, 0, 0, 0, self()]).

handle_PID(I2C, PreviousAngle, PreviousError, ErrorSum, Parrent) ->
    Data = read(I2C),

    AccY = Data#mpu.accy,
    AccZ = Data#mpu.accz,
    GyroX = Data#mpu.gyrox,
    
    AccAngle = math:atan2(AccY, AccZ) * ?RADIAN_TO_DEGREE,
    GyroAngle = GyroX * ?SAMPLE_TIME_S,
    CurrentAngle = ?ALPHA*(PreviousAngle + GyroAngle) + (1 - ?ALPHA)*(AccAngle),

    Kp = 40, Kd = 0.75, Ki = 150,
    
    Error = ?TARGET_ANGLE - CurrentAngle,
    NewErrorSum = constrain(ErrorSum + Error, -400, 400),
    %NewErrorSum = ErrorSum + Error,
    Power = Kp*(Error) + Ki*(NewErrorSum)*?SAMPLE_TIME_S + Kd*(Error-PreviousError)*?DIV_SAMPLE_TIME_S,

    case (abs(CurrentAngle) > 30) of
        true ->
            Parrent ! {stop, CurrentAngle};
        false ->
            Parrent ! {power, Power, CurrentAngle},
            %Parrent ! {angle, CurrentAngle},
            timer:sleep(1),
            handle_PID(I2C, CurrentAngle, Error, NewErrorSum, Parrent)
    end.
    
constrain(Value, Low, High) ->
    if
        Value > High ->
            High;
        Value < Low ->
            Low;
        true ->
            Value
    end. 
    

% Read value from 6500 N times and return record contain result
% read_N_times(_I2C, Data, 0) ->
%     AccError = Data#error.accerror * ?DIV_TIMES,
%     GyroError = Data#error.gyroerror * ?DIV_TIMES,
%     #error{accerror = AccError, gyroerror = GyroError};
    
% read_N_times(I2C, PreviousData, Times) ->
% 	Data = read(I2C),
%     AccError = math:atan2(Data#mpu.accy, Data#mpu.accz) * ?RADIAN_TO_DEGREE,
% 	NewData = #error{
%         accerror = AccError + PreviousData#error.accerror,
% 		gyroerror = Data#mpu.gyrox + PreviousData#error.gyroerror
% 		},
%     usleep(5),
% 	read_N_times(I2C, NewData, Times - 1).

% usleep(Time) when is_integer(Time) andalso Time >= 0 ->
%    do_usleep(Time),
%    receive
%        sleep_done ->
%            ok
%    end.
% do_usleep(Time) ->
%     UsSecs = erlang:system_time(microsecond),
%     do_usleep(UsSecs, Time).

% do_usleep(Start, Time) ->
%     UsSecs = erlang:system_time(microsecond),
%     case UsSecs - Start >= Time of
%         true ->
%             self() ! sleep_done;
%         _ ->
%             do_usleep(Start, Time)
%     end.

map(Value, InLow, InHigh, OutLow, OutHigh) ->
    Res = (Value - InLow) * (OutHigh - OutLow) / (InHigh - InLow) + OutLow,
    round(Res).

File: projects/elixir/self_balance_robot/lib/self_balance_robot.ex
# Copyright (c) 2024 AtomVM
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# This file is part of AtomVM.
#
# Copyright 2026 Bien Nguyen <nguyennhubientdh94@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
#

defmodule SelfBalanceRobot do
  defstruct [:accy, :accz, :gyrox, :temp]

  @ledc_high_speed_mode 0
  @ledc_low_speed_mode 1
  @ledc_timer_0 0
  @ledc_timer_1 1
  @ledc_timer_2 2
  @ledc_timer_3 3
  @ledc_channel_0 0
  @ledc_channel_1 1
  @ledc_channel_2 2
  @ledc_channel_3 3
  @ledc_channel_4 4
  @ledc_channel_5 5
  @ledc_channel_6 6
  @ledc_channel_7 7
  @ledc_timer_1_bit 1
  @ledc_timer_2_bit 2
  @ledc_timer_3_bit 3
  @ledc_timer_4_bit 4
  @ledc_timer_5_bit 5
  @ledc_timer_6_bit 6
  @ledc_timer_7_bit 7
  @ledc_timer_8_bit 8
  @ledc_timer_9_bit 9
  @ledc_timer_10_bit 10
  @ledc_timer_11_bit 11
  @ledc_timer_12_bit 12
  @ledc_timer_13_bit 13
  @ledc_timer_14_bit 14
  @ledc_timer_15_bit 15
  @ledc_timer_16_bit 16
  @ledc_timer_17_bit 17
  @ledc_timer_18_bit 18
  @ledc_timer_19_bit 19
  @ledc_timer_20_bit 20
  @ledc_fade_no_wait 0
  @ledc_fade_wait_done 1

  @ledc_hs_timer @ledc_timer_0
  @ledc_hs_mode @ledc_high_speed_mode
  @ledc_hs_ch0_gpio 19
  @ledc_hs_ch0_channel @ledc_channel_0
  @ledc_hs_ch1_gpio 18
  @ledc_hs_ch1_channel @ledc_channel_1
  @cal_bit_div_100 81.91
  @cal_100_div_bit 0.012208521548040533
  @alpha 0.988
  @sample_time_ms 5
  @sample_time_s 0.01
  @div_sample_time_s 100
  @target_angle -2
  @gpio_scl 22
  @gpio_sda 21
  @gpio_led 2
  @motor_left_1 16
  @motor_left_2 4
  @motor_right_1 5
  @motor_right_2 23
  @mpu9250_addr 0x68
  @accel_yout_h 0x3D
  @acc_config_addr 0x1C
  @gyro_config_addr 0x1B
  @i2c_base_freq 400_000
  @gyro_full_scale_250_dps 0x00
  @gyro_full_scale_2000_dps 0x18
  @acc_full_scale_16_g 0x18
  @temp_offset 0
  @temp_sens 0.003115264797507788
  @acc_scale 6.103515625e-5
  @gyro_scale 0.007633587786259542
  @radian_to_degree 57.2957795
  @num_of_byte 8
  @num_of_times 200
  @div_times 0.02
  @acc_offset 4.5
  @gyro_offset -1.75
  @ay_off -0.00247998047
  @az_off 1.00272071
  @gx_off 376.86
  @duty_equal_zero 5000
  @max_speed 8191

  def start do
    :erlang.system_flag(:schedulers_online, 2)
    i2c = i2c_init()
    mpu_config(i2c)
    gpio_init()
    pwm_init()
    process_init(i2c)
    IO.puts("INIT OK")
    loop()
  end

  defp loop do
    receive do
      {:power, power, angle} ->
        new_power = constrain(power, -8191, 8191)
        duty_none_round = abs(new_power)
        if abs(angle) > 10 do
          duty = map(duty_none_round, 0, 8191, 7300, 8191)
        else
          duty = map(duty_none_round, 0, 8191, 5000, 6000)
        end
        IO.puts("Duty #{duty}")
        if abs(angle) < 1 do
          set_counter_channel_1(0)
          set_counter_channel_2(0)
        else
          if new_power <= 0 do
            car_backward()
            set_counter_channel_2(duty)
            set_counter_channel_1(duty)
          else
            car_forward()
            set_counter_channel_2(duty)
            set_counter_channel_1(duty)
          end
        end
      {:angle, angle} ->
        IO.puts("Current Angle is: #{round(angle)}")
      {:stop, angle} ->
        IO.puts("Angle is: #{angle}")
        car_stop()
        set_counter_channel_1(0)
        set_counter_channel_2(0)
    end
    loop()
  end

  defp read(i2c) do
    {:ok, val} = mpu_read_data(i2c)
    <<accy::integer-signed-16, accz::integer-signed-16,
      temp::16, gyrox::integer-signed-16>> = val
    acc_data_y = (accy - @ay_off) * @acc_scale
    acc_data_z = (accz - @az_off) * @acc_scale
    gyro_data_x = (gyrox - @gx_off) * @gyro_scale
    temp_data = get_temp_value(temp)
    %__MODULE__{accy: acc_data_y, accz: acc_data_z, gyrox: gyro_data_x, temp: temp_data}
  end

  defp get_temp_value(temp) do
    (temp - @temp_offset) * @temp_sens + 21
  end

  defp i2c_init do
    I2C.open([scl: @gpio_scl, sda: @gpio_sda, clock_speed_hz: @i2c_base_freq])
  end

  defp mpu_config(i2c) do
    mpu_send_command(i2c, @acc_config_addr, @acc_full_scale_16_g)
    mpu_send_command(i2c, @gyro_config_addr, @gyro_full_scale_250_dps)
  end

  defp mpu_send_command(i2c, register, command) do
    I2C.begin_transmission(i2c, @mpu9250_addr)
    I2C.write_byte(i2c, register)
    I2C.write_byte(i2c, command)
    I2C.end_transmission(i2c)
  end

  defp mpu_read_data(i2c) do
    I2C.begin_transmission(i2c, @mpu9250_addr)
    I2C.write_byte(i2c, @accel_yout_h)
    I2C.end_transmission(i2c)
    I2C.read_bytes(i2c, @mpu9250_addr, @num_of_byte)
  end

  defp gpio_init do
    list = [@motor_left_1, @motor_left_2, @motor_right_1, @motor_right_2, @gpio_led]
    Enum.each(list, fn gpio -> GPIO.set_pin_mode(gpio, :output) end)
    car_stop()
  end

  defp write_motor_left(status1, status2) do
    GPIO.digital_write(@motor_left_1, status1)
    GPIO.digital_write(@motor_left_2, status2)
  end

  defp write_motor_right(status1, status2) do
    GPIO.digital_write(@motor_right_1, status1)
    GPIO.digital_write(@motor_right_2, status2)
  end

  defp car_stop do
    write_motor_left(:low, :low)
    write_motor_right(:low, :low)
  end

  defp car_forward do
    write_motor_left(:low, :high)
    write_motor_right(:low, :high)
  end

  defp car_backward do
    write_motor_left(:high, :low)
    write_motor_right(:high, :low)
  end

  defp pwm_init do
    ledc_hs_timer = [
      duty_resolution: @ledc_timer_13_bit,
      freq_hz: 5000,
      speed_mode: @ledc_hs_mode,
      timer_num: @ledc_hs_timer
    ]
    :ok = :ledc.timer_config(ledc_hs_timer)
    ledc_channel_1 = [
      channel: @ledc_hs_ch0_channel,
      duty: 0,
      gpio_num: @ledc_hs_ch0_gpio,
      speed_mode: @ledc_hs_mode,
      hpoint: 0,
      timer_sel: @ledc_hs_timer
    ]
    ledc_channel_2 = [
      channel: @ledc_hs_ch1_channel,
      duty: 0,
      gpio_num: @ledc_hs_ch1_gpio,
      speed_mode: @ledc_hs_mode,
      hpoint: 0,
      timer_sel: @ledc_hs_timer
    ]
    :ok = :ledc.channel_config(ledc_channel_1)
    :ok = :ledc.channel_config(ledc_channel_2)
    :ok = :ledc.fade_func_install(0)
  end

  defp set_counter_channel_1(duty_num) do
    speed_mode = @ledc_hs_mode
    channel = @ledc_hs_ch0_channel
    duty = round(duty_num)
    :ok = :ledc.set_duty(speed_mode, channel, duty)
    :ok = :ledc.update_duty(speed_mode, channel)
  end

  defp set_counter_channel_2(duty_num) do
    speed_mode = @ledc_hs_mode
    channel = @ledc_hs_ch1_channel
    duty = round(duty_num)
    :ok = :ledc.set_duty(speed_mode, channel, duty)
    :ok = :ledc.update_duty(speed_mode, channel)
  end

  defp process_init(i2c) do
    spawn(fn -> handle_pid(i2c, 0, 0, 0, self()) end)
  end

  def handle_pid(i2c, previous_angle, previous_error, error_sum, parent) do
    data = read(i2c)
    acc_y = data.accy
    acc_z = data.accz
    gyro_x = data.gyrox

    acc_angle = :math.atan2(acc_y, acc_z) * @radian_to_degree
    gyro_angle = gyro_x * @sample_time_s
    current_angle = @alpha * (previous_angle + gyro_angle) + (1 - @alpha) * acc_angle

    kp = 40
    kd = 0.75
    ki = 150

    error = @target_angle - current_angle
    new_error_sum = constrain(error_sum + error, -400, 400)
    power = kp * error + ki * new_error_sum * @sample_time_s + kd * (error - previous_error) * @div_sample_time_s

    case abs(current_angle) > 30 do
      true ->
        send(parent, {:stop, current_angle})
      false ->
        send(parent, {:power, power, current_angle})
        Process.sleep(1)
        handle_pid(i2c, current_angle, error, new_error_sum, parent)
    end
  end

  defp constrain(value, low, high) do
    cond do
      value > high -> high
      value < low -> low
      true -> value
    end
  end

  defp map(value, in_low, in_high, out_low, out_high) do
    res = (value - in_low) * (out_high - out_low) / (in_high - in_low) + out_low
    round(res)
  end
end
